pylabnet/hardware/ZI_HDAWG/zi_hdawg.py

```diff
@@ -0,0 +1,475 @@
+# -*- coding: utf-8 -*-
+
+"""
+This file contains the pylabnet Hardware class
+for the Zurich Instruments HDAWG.
+"""
+
+import zhinst.utils
+import re
+import time
+import textwrap
+import copy
+
+from pylabnet.utils.logging.logger import LogHandler
+
+from pylabnet.utils.decorators.logging_redirector import log_standard_output
+
+
+class HDAWGDriver():
+
+    def disable_everything(self):
+        """ Create a base configuration.
+        Disable all available outputs, awgs, demods, scopes, etc.
+        """
+        zhinst.utils.disable_everything(self.daq, self.device_id)
+        self.log.info("Disabled all wave outputs.")
+
+    @log_standard_output
+    def log_stdout(self, function):
+        """ Execute function and log print output to self.log
+
+        This statement is needed for an inline call where any zhinst command is
+        executed and the standard output should be logged
+        :function: The function to be executed.
+         """
+        return function()
+
+    def _convert_to_list(self, input_argument):
+        """Checks if input is list and if not, converts it to list."""
+        if type(input_argument) is not list:
+            input_argument = [input_argument]
+        return input_argument
+
+    def __init__(self, device_id, logger, api_level=6):
+        """ Instantiate AWG
+
+        :logger: instance of LogClient class
+        :device_id: Device id of connceted ZI HDAWG, for example 'dev8060'
+        :api_level: API level of zhins API
+        """
+
+        # Instantiate log
+        self.log = LogHandler(logger=logger)
+
+        # Part of this code has been modified from
+        # ZI's Zurich Instruments LabOne Python API Example
+
+        # Call a zhinst utility function that returns:
+        # - an API session `daq` in order to communicate
+        # with devices via the data server.
+        # - the device ID string that specifies the device
+        # branch in the server's node hierarchy.
+        # - the device's discovery properties.
+
+        err_msg = "This example can only be run on an HDAWG."
+
+        # Connect to device and log print output, not the lambda expression.
+        (daq, device, props) = self.log_stdout(
+            lambda: zhinst.utils.create_api_session(
+                device_id,
+                api_level,
+                required_devtype='HDAWG',
+                required_err_msg=err_msg
+            )
+        )
+
+        self.log_stdout(lambda: zhinst.utils.api_server_version_check(daq))
+
+        self.daq = daq
+        self.device_id = device
+
+        # Create a base configuration
+        self.disable_everything()
+
+        # read out number of channels from property dictionary
+        self.num_outputs = int(
+            re.compile('HDAWG(4|8{1})').match(props['devicetype']).group(1)
+        )
+
+    @log_standard_output
+    def seti(self, node, new_int):
```

Comment:
The `HDAWG_Driver` class is just our usual `pylabnet` hardware class. The `AWGModule` class should be a wrapper of the `zhinst`-native `daq.awgModule()`.  This module, somewhat weirdly named, only deals with the execution of a sequence of one AWG core. The reason why I added the `AWGModule` class is to access the `daq.awgModule()` in a more convenient way and to easily tie it to the `HDAWG_Driver`. So the logic is:

`HDAWG_Driver`: Acceses all general purpose HDAWG device functions
 `AWGModule`: Accesses all functions related to uploading sequences to an AWG core and executing said sequence.

Now that I look at it I can see why the distinction might not be super clear. I'm open to suggestions how to partition this differently.

Reply:
i see - it makes sense since in principle one can use/program different cores completely independently. i think it sounds reasonable, although I'm still not 100% sure I fully understand which functionality belongs where (for example, why is the output range part of the Driver, and not the Module?). For now, I'm ok proceeding and reevaluating at a later time if necessary

Reply:
I think this separation is tied to the fact how ZI has implemented LabOne. A (unit-less) waveform as defined in the sequence should always be in the range from -1 to 1. This number then will get multiplied with the output range, which is set in the "Wave Output" node and then fed into the digital to analogue converter.